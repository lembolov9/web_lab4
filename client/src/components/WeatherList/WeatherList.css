.WeatherList {
    display: grid;
    grid-template-columns: 1fr 1fr;
}

.loader {
    margin: auto;
    width: 6em;
    height: 6em;
    border: 1.1em solid rgba(0, 0, 0, 0.2);
    border-left: 1.1em solid lavender;
    border-radius: 50%;
    animation: load8 1.1s infinite linear;
    transition: opacity 0.3s;
}

@keyframes load8 {
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
    }
}