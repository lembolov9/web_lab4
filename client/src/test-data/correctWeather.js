export default  {
    "coord": {
        "lon": 30.52,
        "lat": 50.43
    },
    "weather": [
        {
            "id": 500,
            "main": "Rain",
            "description": "light rain",
            "icon": "10n"
        }
    ],
    "base": "stations",
    "main": {
        "temp": 8.02,
        "pressure": 1011,
        "humidity": 100,
        "temp_min": 7.78,
        "temp_max": 8.33
    },
    "visibility": 6000,
    "wind": {
        "speed": 3,
        "deg": 80
    },
    "rain": {
        "1h": 0.76
    },
    "clouds": {
        "all": 90
    },
    "dt": 1573154160,
    "sys": {
        "type": 1,
        "id": 8903,
        "country": "UA",
        "sunrise": 1573102677,
        "sunset": 1573136713
    },
    "timezone": 7200,
    "id": 703448,
    "name": "Kiev",
    "cod": 200
};