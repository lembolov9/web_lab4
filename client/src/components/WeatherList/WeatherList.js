import React from "react";
import Weather from "../Weather/Weather";
import {
    getWeatherWait,
    getWeatherError,
    getWeatherSuccess,
    deleteCity,
    deleteCitySuccess,
    deleteCityError, getCitiesWait, getCitiesSuccess, getCitiesError
} from '../../actions';
import { connect } from 'react-redux';
import axios from "axios"
import {getCities, getCityForDelete, getError, getIsGetting, getNewCity} from "../../reducers"
import "./WeatherList.css"
import {api} from "../../constants/server";

export const getWeatherByNameAction = (id, cityName) => {
    return dispatch => {
        dispatch(getWeatherWait(id));
        axios
            .get(`${api}weather?city=${cityName}`)
            .then(response => {
                dispatch(getWeatherSuccess(response.data, id));
                console.log(response.data);
            })
            .catch(error => {
                dispatch(getWeatherError(error, id));
                console.log(error);
            });
    }
};

const deleteCityAction = id => {
    return dispatch => {
        dispatch(deleteCity(id));
        axios
            .delete(`${api}favourites/${id}`)
            .then(response => {
                dispatch(deleteCitySuccess(id));
            })
            .catch(error => {
                dispatch(deleteCityError(error));
            });
    };
};

const getCitiesAction = () => {
    return dispatch => {
        dispatch(getCitiesWait());
        axios
            .get(`${api}favourites`)
            .then(response => {
                const cities = response.data.map(city => ({ name: city.name, id: city._id }));
                dispatch(getCitiesSuccess(cities));
            })
            .catch(error => {
                dispatch(getCitiesError(error));
            });
    };
};

const mapStateToProps = (state) => ({
    cities: getCities(state),
    error: getError(state),
    newCityId: getNewCity(state),
    isGetting: getIsGetting(state),
    cityToDelete: getCityForDelete(state)
});

const mapDispatchToProps = (dispatch) => ({
    getCities: () => dispatch(getCitiesAction()),
    getWeatherByName: (id, name) => dispatch(getWeatherByNameAction(id, name)),
    deleteCity: id => dispatch(deleteCityAction(id))
});


class WeatherList extends React.Component {

    componentDidMount(){
        this.props.getCities();
    }

    render() {
        if (this.props.error !== null) {
            if (this.props.isGetting) {
                return (
                    <div className="WeatherList">
                        Error with get cities'
                    </div>
                );
            }

            if (this.props.cityForDelete !== null) {
                alert('Cannot delete city');
            }

            if (this.props.newCity !== null) {
                alert('Cannot add new city');
            }

        } else {
            if (this.props.isGetting) {
                return (
                    <div className="WeatherList">
                        <div className="loader"/>
                    </div>
                );
            }
        }
        if (Array.isArray(this.props.cities)) {
            return (
                <div className="WeatherList">
                    {this.props.cities.map(city => (
                        <Weather key={city.id} data={city}
                                 getWeather={() => this.props.getWeatherByName(city.id, city.name)}
                                 deleteCity={() => this.props.deleteCity(city.id)}/>
                    ))}
                </div>
            );
        } else {
            return (
                <div className="WeatherList"/>
            )
        }
    }
}

const WList = connect(mapStateToProps, mapDispatchToProps)(WeatherList);


export {WList , WeatherList};