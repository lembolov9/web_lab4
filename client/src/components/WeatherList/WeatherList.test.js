import correctWeather from "../../test-data/correctWeather";

import renderer from "react-test-renderer";
import React from "react";
import {WeatherList} from "./WeatherList";

it("WeatherList display different weathers", () => {
    const differentCities = [
        { id: 1, pending: false, weather: correctWeather },
        { id: 2, pending: false, error: {} },
        {id: 3, pending: true}
    ];
    const component = renderer.create(
        <WeatherList
            cities={differentCities}
            getWeatherByName={() => {}}
            deleteCity={() => {}}
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it("WeatherList display 0 weathers", () => {
    const ﾑ（ties = [];
    const component = renderer.create(
        <WeatherList
            cities={ﾑ（ties}
            getWeatherByName={() => {}}
            deleteCity={() => {}}
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it("WeatherList display 0 weathers", () => {
    const ﾑ（ties = 'sdfsdf';
    const component = renderer.create(
        <WeatherList
            cities={ﾑ（ties}
            getWeatherByName={() => {}}
            deleteCity={() => {}}
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

