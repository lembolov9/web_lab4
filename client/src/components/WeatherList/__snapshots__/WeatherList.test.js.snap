// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherList display 0 weathers 1`] = `
<div
  className="WeatherList"
/>
`;

exports[`WeatherList display 0 weathers 2`] = `
<div
  className="WeatherList"
>
  Error, please clean localstorage
</div>
`;

exports[`WeatherList display different weathers 1`] = `
<div
  className="WeatherList"
>
  <div
    className="Weather"
  >
    <div
      className="loader"
    />
  </div>
  <div
    className="Weather error"
  >
    <div>
      Shit happens
    </div>
    <button
      className="circle"
      onClick={[Function]}
    >
      x
    </button>
  </div>
  <div
    className="Weather"
  >
    <div
      className="loader"
    />
  </div>
</div>
`;
